src/client/client.cpp

```diff
@@ -547,7 +548,9 @@ void Client::step(float dtime)
 		int num_processed_meshes = 0;
 		std::vector<v3s16> blocks_to_ack;
 		bool force_update_shadows = false;
-		while (!m_mesh_update_thread.m_queue_out.empty())
+		auto mesh_queue_size = m_mesh_update_thread.m_queue_out.size();
+		m_avg_mesh_queue_size = MYMIN(mesh_queue_size, 0.95 * m_avg_mesh_queue_size + 0.05 * mesh_queue_size);
```

Comment:
m_avg_mesh_queue_size is an int and starts with 0.

The result of this is cast to an int. If mesh_queue_size is not at least 20 at least a few time, m_avg_mesh_queue_size will remain 0, right?

Either round, or make m_avg_mesh_queue_size a float.

Reply:
I'm testing with larger blocks, so presumably my queue size are lower. This is still a problem, though.

Reply:
Also, doesn't this defeat the purpose to some extend?

First we speed up mesh generation with multiple threads, and then we slow down replacing the meshes.
I preferred the simpler approach without this spreading.

Reply:
Fixes applied, and I tuned the adaptive behavior to reduce the lag.

No, it does not. Threads increase the processing capacity, but this limit controls the lag per-frame, reducing stutter. Changing the number of threads is too coarse and does not solve the problem of mapblocks arriving in groups.

With the latest changes the system will aim at 1, at most 2-frame lag for the meshes, plus all the 'urgent' ones are at the beginning of the queue, further reducing their latency.